### DISCUSSION TITLE
Custom Script get VLAN of a Interface - solved

### DISCUSSION THREAD

**Discussion by do5tj80** [netbox-community/netbox#10534] [General]
Hi,
I would like to read out the VLAN ID of the interfaces for all devices (depending on the tag) through a custem script.
Unfortunately, I am not able to do this at the moment.
Can anyone give me a tip on how best to get the desired data?

Many thanks 

Tobias

**Comment by candlerb** [comment]
When writing a custom script, you're talking directly to the Django ORM. [nbshell](https://docs.netbox.dev/en/stable/administration/netbox-shell/) (the netbox shell) is the best way to explore this.

```
cd /opt/netbox
. venv/bin/activate
python netbox/manage.py nbshell
```

There are some sample scripts [here](https://github.com/netbox-community/reports/tree/master/scripts), but otherwise it's up to you to learn how to use python and the Django ORM.

If you've explicitly configured VLANs on each interface (e.g. set mode "access" or "tagged") then this information is available directly as properties on the interface in question.  Example from nbshell:

```
>>> d = Device.objects.get(name="gw1")
>>> d.interfaces
<django.db.models.fields.related_descriptors.create_reverse_many_to_one_manager.<locals>.RelatedManager object at 0x7f38aff93d00>
>>> d.interfaces.all()
<RestrictedQuerySet [<Interface: ether1>, <Interface: ether2>, <Interface: ether3>, <Interface: ether4>, <Interface: ether5>, <Interface: pppoe-spare2>, <Interface: sfp1>, <Interface: sit1>, <Interface: vlan250>, <Interface: vlan254>]>
>>> i = d.interfaces.all()[4]
>>> i
<Interface: ether5>
>>> help(Interface)
     ### shows you all the help information and properties ###
>>> i.untagged_vlan
<VLAN: mgmt-dmz (255)>
>>> i.tagged_vlans
<django.db.models.fields.related_descriptors.create_forward_many_to_many_manager.<locals>.ManyRelatedManager object at 0x7f38aff93e50>
>>> i.tagged_vlans.all()
<VLANQuerySet [<VLAN: nsrc-alt (251)>, <VLAN: nsrc-lan (252)>, <VLAN: family (254)>, <VLAN: admin (256)>]>
```

If you haven't configured VLANs on each interface, then you'll need to enumerate all the IP addresses on the interface, find out which prefix each one belongs to, and see what VLAN each prefix belongs to. This is similar to discussion #10457, except that was using the REST API rather than the Django ORM.

**Comment by do5tj80** [comment]
Thank you Brian, that was the clue I was missing.
I have all the data I need.
